### ISSUE TITLE
Create a ROS 2 linter integration

### ISSUE THREAD

**Issue by gavanderhoorn** [rosin-project/haros_catkin#24]
Would/Should not end up in this repository, but a good thing to initiate/track on this tracker.

For ROS 2 integration we'd need to create a new linter package.

Lifting @nlimpert's comment from https://github.com/rosin-project/haros_catkin/issues/14#issuecomment-526118974:

> [This](https://ubuntu.com/blog/how-to-add-a-linter-to-ros-2) seems to be a nice introduction which looks like it boils down to:
> 
> 1. Create a ROS package to contain the linter including `setup.py` and the appropriate python script to execute the linter and return an appropriate value. This can for example also be an invocation of `clang-format` - see [here](https://github.com/ament/ament_lint/tree/cc1000824237cd74dec3e0ff210ee08020c2cd92/ament_clang_format/ament_clang_format)).
>        To adhere to Jenkins' ability to parse output the script has to generate JUnit XML output (this is probably of interest in terms of @max-krichenbauer's [suggestion](https://github.com/ros-infrastructure/ros_buildfarm/issues/596#issuecomment-526015764) to display HAROS output on the ROS Wiki as well as on Jenkins.
> 
> 2. Create another package that contains the appropriate CMake macros to invoke the script created in 1.
> 
> 3. The package created in 2. has to define a set of dependencies in its `package.xml` to adhere to `ament`'s registration for test setups. It also needs the CMake Configuration to invoke `ament_copyright()` and `ament_lint_cmake()`. This package also has to provide a CMake macro to invoke the CMake macro created in 1.
> 
>  4. The actual CMake macro that invokes the newly introduced linter looks like [this](https://github.com/ament/ament_lint/blob/cc1000824237cd74dec3e0ff210ee08020c2cd92/ament_cmake_clang_format/cmake/ament_clang_format.cmake).
> 
> 
> So the overall procedure generally seems to take two packages - one for the actual linter process (being a CLI wrapper) and another one to contain the CMake scripts to register the linter to run during tests.

**Comment by max-krichenbauer** [comment]
Thank you!
For now I have started to work on adding JUnit XML format output for HAROS, since it relates to the other topic.

**Comment by max-krichenbauer** [comment]
Here is the PR for a first implementation of JUnit XML format output:
https://github.com/git-afsantos/haros/pull/79

**Comment by max-krichenbauer** [comment]
Reading through the tutorial @nlimpert shared, I found two issues that I'd like to ask about:

1) it needs capabilities for excluding directories with ``AMENT_IGNORE``
I don't think HAROS has this yet. I can offer making a PR to HAROS to add it as a feature.

2) ~~HAROS needs to be in [ROSDistro](https://github.com/ros/rosdistro) so it can be a dependency for the package. Has is already been submitted? Is anyone already working on it?~~

**EDIT**: I think the tutorial is incorrect (or at least misleading) in this point. 
[``ament_cppcheck``](https://github.com/ament/ament_lint/blob/master/ament_cppcheck/package.xml) specifies a ``<exec_depend>cppcheck</exec_depend>`` which is not in ROSdistro,
and even [``ament_mypy``](https://github.com/ament/ament_lint/blob/master/ament_mypy/package.xml) (on which the tutorial is based) specifies ``<exec_depend>python3-mypy</exec_depend>``, not in ROSdistro.

**Comment by gavanderhoorn** [comment]
I believe you are confusing `distribution.yaml` with the `rosdep` db.

`distribution.yaml` is only for ROS packages, which `cppcheck` is not. `cppcheck` does have [a rosdep key though](https://github.com/ros/rosdistro/blob/6c20fd7b4dfaf5878cf53ba4d70f0e7bec202eb6/rosdep/base.yaml#L519-L527), which is sufficient for `rosdep` to be able to install it when a ROS package depends on it.

So Haros would have to:

 - either be a ROS package, or
 - have a rosdep key

For `haros_catkin` we've gone with the first option. See https://github.com/ros/rosdistro/pull/22134 and https://github.com/ros/rosdistro/pull/22146.

**Comment by max-krichenbauer** [comment]
Ahh, thanks - that clears things up!

In the meantime, I've written the ``ament_haros`` and ``ament_cmake_haros`` packages for local testing, but have run into several issues with HAROS there.
Basically, it seems like HAROS can't handle ROS2 workspaces (relying on ``ROS_WORKSPACE``, not finding ``CMakeLists.txt`` correctly, not respecting ``AMENT_IGNORE`` ...).

Is it intended to add ROS2 support to HAROS?

**Comment by gavanderhoorn** [comment]
> Is it intended to add ROS2 support to HAROS?

https://github.com/git-afsantos/haros/issues/32.

> not respecting `AMENT_IGNORE` ...

I don't think this is something that should be done by the tool. We cannot expect every tool to understand ROS workspace semantics, so this seems like a strange requirement.

Haros is of course somewhat different, in that its purpose is to parse and work with ROS workspaces, but even then we might want to see whether it'd be better to delegate it to the integration to come up with a list of paths to inspect for Haros.

Also: Haros (at least currently) does not require a workspace to have been built, as it performs a static analysis. The only thing it really needs is to be able to find packages.

**Comment by max-krichenbauer** [comment]
I see. Thank you so much for your explanation!

So could a next logical step then be a PR to HAROS to add a feature manually select the CMakeList.txt files?
Then ``ament_haros`` (or ``haros_catkin``) could take care of the parsing and selection process.

**Comment by gavanderhoorn** [comment]
We might benefit from some input from @git-afsantos here.

Haros does not only parse `CMakeLists.txt`. Perhaps getting it to parse a certain list of directories might make sense, but lets see what @git-afsantos thinks.

**Comment by git-afsantos** [comment]
I think I might need some more context. What is the purpose of `AMENT_IGNORE`? Is it to ignore packages within a workspace, or to ignore directories within a package?

> Basically, it seems like HAROS can't handle ROS2 workspaces

True. I have not worked with (or aiming for) ROS2 yet, neither will I until I conclude my PhD.

@gavanderhoorn 
> getting it to parse a certain list of directories might make sense

What do you mean? Do you mean providing Haros the list of directories that is supposed to contain packages (and then let Haros find them), or something more specific?

**Comment by gavanderhoorn** [comment]
> What is the purpose of `AMENT_IGNORE`? Is it to ignore packages within a workspace, or to ignore directories within a package?

It's essentially the `CATKIN_IGNORE` of Ament. But to make it confusing, Ament also 'respects' `CATKIN_IGNORE`s ;)

As to what it does: it's slightly more nuanced: encountering a `(CATKIN|AMENT)_IGNORE` will cause that directory (and all its sub directories) to be ignored, completely. So not just packages, but everything.

If you place such a file in a directory containing a package, the result will be that the package will not be considered part of the workspace. If it's placed in a sub directory of a package's directory, the package will be found, but if there would be 'sub packages' in that sub directory (they don't exist, but it's just an example), or other files recognised by either Catkin or Ament, those would be ignored.

> > getting it to parse a certain list of directories might make sense
> 
> What do you mean? Do you mean providing Haros the list of directories that is supposed to contain packages (and then let Haros find them), [..]?

Yes, this.

**Comment by git-afsantos** [comment]
> It's essentially the `CATKIN_IGNORE` of Ament. But to make it confusing, Ament also 'respects' `CATKIN_IGNORE`s ;)

Ok. I do not think Haros even respects the catkin version, so it might make for a good addition.

> [...] not just packages, but everything

So this implies two filters. Discarding packages within such directories, and discarding source files as well (e.g. `AMENT_IGNORE` within `pkg/src/test_code`).

I think Haros only ignores a few hard-coded directories currently (`.git`, `doc`, `bin`, ...). So, any `*_IGNORE` directory would have to be added to this list.

> providing Haros the list of directories that is supposed to contain packages

Haros currently has an option to look for packages within (and below) a given directory (or the cwd, cannot remember). This is probably a good starting point, since what you want to do is basically go from N=1 to N=*, and I would be ok with that as a PR.

Another alternative, so as not to make commands extremely long, is to extend either the project YAML files or the settings YAML file with such a list of dirs.

**Comment by max-krichenbauer** [comment]
Thank you for the comments!

As far as I understand, you would not need both features though, right?

If HAROS respects ``(CATKIN|AMENT)_IGNORE`` and understands the ROS2 workspace structure, there's no need to supply the list of files and folders manually.

If you supply the list of folders manually, whichever tool inputs the files/folders can take care of that.

Personally, I would go with the first option (updating HAROS's parser). HAROS is ROS specific to the point that it's even in the name of the tool, and I think ROS2 users would prefer to use just HAROS directly, without any pre-processing-parsing tool.

**EDIT**: I looked into the current version of the HAROS parser and found that it relies on ``rospkg.RosPack``. This may cause a problem (or: is currently causing problems) with ROS2 workspaces.
Not sure if ``rospkg`` is expected to be replaced or if it just isn't "ROS2-ready" yet.

**Comment by git-afsantos** [comment]
> If HAROS respects `(CATKIN|AMENT)_IGNORE` and understands the ROS2 workspace structure, there's no need to supply the list of files and folders manually.

Yes, I believe that is correct.

> Personally, I would go with the first option (updating HAROS's parser).

This is my preference as well. :+1:
For me, HAROS was always intended to understand ROS-specific details. That is what sets it apart from a general-purpose tool.

**Comment by max-krichenbauer** [comment]
Thank you for the clarification!

I understand that you're currently busy with your PhD.
I might have time to work on this (and other issues) this month, but it would be important to know that such a contribution is actually welcome, going in the right direction, and has a chance to be merged in the near future.
If you rather want to do it yourself I can understand that as well.
Please let me know what you think.

**Comment by git-afsantos** [comment]
@max-krichenbauer Indeed, I am now nearing the final months/semester of my PhD. During this time, it is likely that I will not pick this up for implementation. If you want to go ahead and try it, you are most welcome to do so. :+1:

**Comment by max-krichenbauer** [comment]
Thanks and good luck with your graduation!
When is the graduation planned?

Then I will see if I can patch together at least minimal support for ROS2.
If I make a PR, would you consider it?

**Comment by git-afsantos** [comment]
Thanks, @max-krichenbauer! We do not have a precise date yet, as I am still about to finish my last *implementation work* (property-based testing and runtime monitoring), and then I have to write it all down. I expect early 2020.

> Then I will see if I can patch together at least minimal support for ROS2.
> If I make a PR, would you consider it?

Sure, that would be great! Even when I do not implement things myself, I always try to fix bugs and review PRs as soon as possible.

**Comment by max-krichenbauer** [comment]
> > Then I will see if I can patch together at least minimal support for ROS2.
> > If I make a PR, would you consider it?
> 
> Sure, that would be great! Even when I do not implement things myself, I always try to fix bugs and review PRs as soon as possible.

Thanks! Happy to hear that. Then I'll get to work on it.
In the meantime, could you please have a look at https://github.com/git-afsantos/haros/pull/79 and let me know if there's any problem with it?

**Comment by max-krichenbauer** [comment]
I started working on adding compatibility for ROS2 workspaces here:
https://github.com/esol-community/haros/commit/6dcb4e15d4b230e339f2e78590ec4996f0f9b345

One issue I'm having is that pure Python ROS2 packages don't necessarily seem to have a ``CMakeLists.txt`` file. Then, the Python files also don't show up in the analysis result.
I'd be happy for any input on the topic.

**Comment by gavanderhoorn** [comment]
My first suggestion would be to not try to merge ROS 1 and ROS 2 workspace handling, as I believe you are trying to do now (removal of `rospkg` use and similar changes).

From a quick look it seems that would introduce a lot of new code, which seems like an anti-pattern to me, as that would seem to duplicate functionality already present in `rospkg` and other similar libraries available in ROS 1.

**Comment by max-krichenbauer** [comment]
``rospkg`` doesn't support ROS2 and I don't know of any ``rospkg2`` development, so at least for now it's up to us to provide that functionality.
https://github.com/ros-infrastructure/rospkg/issues/174

AFAIK ROS2 (colcon) can also handle catkin and there is some degree of ROS1-ROS2 compatibility/interoperability.

So even if you see this as preparation for a "HAROS2" version for ROS2, I think the approach would be the same.

**Comment by gavanderhoorn** [comment]
> `rospkg` doesn't support ROS2 and I don't know of any `rospkg2` development, so at least for now it's up to us to provide that functionality.

my suggestion was to factor out the ROS 2 specific parts into a separate file, and use runtime detection and/or a command line option to force Haros to expect a specific version.

> AFAIK ROS2 (colcon) can also handle catkin and there is some degree of ROS1-ROS2 compatibility/interoperability.

Colcon can build ROS 1 packages, yes. That does not mean we should merge everything into on big `if-else` tree though.

> So even if you see this as preparation for a "HAROS2" version for ROS2, I think the approach would be the same.

My point was that it seems like you're removing `rospkg` calls, and then replacing them with an equivalent piece of code that you wrote yourself. That seems inefficient.

**Comment by max-krichenbauer** [comment]
I see your point.
My worry is that ``rospgk`` is actually detecting ROS1 workspaces within ROS2 workspace temporary / build files. And since HAROS is only using one single function from RosPack I thought it reasonable to provide a ROS1/ROS2 aware version here.
Of course we can also add more command line options to be explicit.

**Comment by max-krichenbauer** [comment]
Here is the ``ament_haros`` ROS2 linter integration package that I'm working on.
https://github.com/esol-community/ament_lint/tree/ament_haros
It's not finished and obviously contains some work-arounds, but the basic concept seems to be working.

**Comment by max-krichenbauer** [comment]
@gavanderhoorn 
I updated the code again: in ROS1 environments, ``rospgk`` is again used, in ROS2 environments, ``colcon`` is used. This way there is no code duplication.
I created a PR here: https://github.com/git-afsantos/haros/pull/83

**Comment by max-krichenbauer** [comment]
The ROS2 linter integration ``ament_haros`` is finished.
https://github.com/esol-community/ament_lint/tree/ament_haros

### How to use:

**1: Download, build and set up environment:**
_(Should not be necessary anymore if it makes it into the official repo._

    # Install haros and haros-plugins:
    pip install haros haros-plugins
    
    # In your ROS2 workspace:
    git clone -b ament_haros --depth 1  https://github.com/esol-community/ament_lint.git
    colcon build --packages-up-to ament_cmake_haros
    . install/setup.bash

**2: Set up your package to use ``ament_haros``**
In your package's CMakeLists.txt add:

    if(BUILD_TESTING)
        find_package(ament_cmake_haros REQUIRED)
        ament_haros()
    endif()

**3: Execution:**

    colcon build --packages-up-to my_package
    colcon test --package-select my_package
    concon test-result

The resulting jUnit XML file will be at ``build/my_package/test_results/my_package/haros.xunit.xml``.

Since https://github.com/git-afsantos/haros/pull/83 and https://github.com/git-afsantos/haros/pull/84 are not merged into HAROS yet, it's still using a manual download from our gitub fork.
I hope we can solve this soon and then send a PR to include ``ament_haros`` as ROS2 linter in the official repo.

I know you're very busy with other things, but it would be great to get your input.

Best regards,
Max

**Comment by git-afsantos** [comment]
@max-krichenbauer Thanks for the contribution! I do not really have any experience with ROS 2, so I cannot comment much on that, but the steps you provide seem simple enough.

I have merged git-afsantos/haros#84, and git-afsantos/haros#83 is just waiting on a couple of comments.

**Comment by gavanderhoorn** [comment]
@max-krichenbauer: nice. This looks good. I'll try to take a look later this week.

:+1:

**Comment by nlimpert** [comment]
> The ROS2 linter integration `ament_haros` is finished.
> https://github.com/esol-community/ament_lint/tree/ament_haros


Hi @max-krichenbauer - thanks for your work! I didn't have the time yet to try out your setup - I'm building ROS2 right now to try it out though.

I was thinking that it might be applicable to introduce a testing package like [`haros_catkin_test`](https://github.com/ipa-jfh/haros_catkin_test) for your work in ROS2. The `haros_catkin` package is used to check `haros_catkin`'s functionality by invoking `haros` when `make` is executed for the `haros_report` target.

So we might just create a new repo evaluating `ament_haros` in a testing package (this might also be used for CI purposes as introduced in https://github.com/rosin-project/haros_catkin/pull/25. What do you think?

**Comment by max-krichenbauer** [comment]
Thank you for your comment @nlimpert 
I'm not familiar with CI integration yet and just started reading reading the documentation at https://github.com/ros-industrial/industrial_ci
but for now I put together a quick test package at [esol-community/ament_lint/ament_haros_test](https://github.com/esol-community/ament_lint/tree/ament_haros/ament_haros_test) which builds and can be tested through colcon with ``ament_haros``.

Could you let me know if this is what you meant and which other files are required in a ROS2 CI context? In ``haros_catkin`` I see two more files ``haros_test.rosinstall`` and ``launch/test.launch``, but I'm not sure who's using those files.

**Comment by nlimpert** [comment]
Hi @max-krichenbauer,

> Could you let me know if this is what you meant and which other files are required in a ROS2 CI context? In `haros_catkin` I see two more files `haros_test.rosinstall` and `launch/test.launch`, but I'm not sure who's using those files.

I'd consider the two files not needed for now so I guess your package `ament_haros_test` looks good! However I wouldn't consider this to also be part of the `ament_lint` repo but another external repository.

In turn I wonder whether the packages `ament_haros` and `ament_cmake_haros` are also needed in the `ament_lint` repo (sorry if this input comes kind of late - if we'd prefer this to be outside of `ament_lint` I'll take the effort to commit this into a separate repo). @gavanderhoorn what do you think? Should we rather create another repo for `ament_haros`?
The latter would also allow us to setup travis CI using `industrial_ci` similar to what we did for `haros_catkin` - without setting up a travis configuration for `ament_lint`.

I can take a look into creating another CI configuration afterwards (similar to `haros_catkin`), but it's most probably not going to differ too much from https://github.com/rosin-project/haros_catkin/commit/d04963766f76bb29be732f7cf91a53920bdc1d9b.

------

I've just tried to run the tests on the ament_haros_test based on your previous comment:

```
colcon build --packages-select ament_haros_test
colcon test --packages-select ament_haros_test
colcon test-result
```
This does, indeed, generate a `build/ament_haros_test/test_results/my_package/haros.xunit.xml` but the contents look like the following:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="ament_haros_test" tests="1" failures="1" time="0" errors="0" skip="0">
  <testcase classname="ament_haros_test" name="haros.xunit.missing_result" status="run" time="0">
    <failure message="The test did not generate a result file:&#10;&#10;Traceback (most recent call last):&#10;  File &quot;/home/nlimpert/dev/ros2_testing_ws/install/ament_haros/bin/ament_haros&quot;, line 11, in &lt;module&gt;&#10;    load_entry_point('ament-haros', 'console_scripts', 'ament_haros')()&#10;  File &quot;/home/nlimpert/dev/ros2_testing_ws/build/ament_haros/ament_haros/main.py&quot;, line 197, in main&#10;    tree = ElementTree(None, haros_tmp_dir + '/haros_data/data/ament_haros_project/compliance/ament_haros_project.xml')&#10;  File &quot;/usr/lib64/python3.7/xml/etree/ElementTree.py&quot;, line 558, in __init__&#10;    self.parse(file)&#10;  File &quot;/usr/lib64/python3.7/xml/etree/ElementTree.py&quot;, line 587, in parse&#10;    source = open(source, &quot;rb&quot;)&#10;FileNotFoundError: [Errno 2] No such file or directory: '/tmp/ament_haros/haros_data/data/ament_haros_project/compliance/ament_haros_project.xml'&#10;"/>
  </testcase>
</testsuite>
```

Note the `FileNotFoundError: [Errno 2] No such file or directory: '/tmp/ament_haros/haros_data/data/ament_haros_project/compliance/ament_haros_project.xml` - is that as it is supposed to be or am I missing something?

Note that I am on Fedora 29 and built ROS Dashing myself.

**Comment by max-krichenbauer** [comment]
Thanks @nlimpert for checking it out.

> In turn I wonder whether the packages `ament_haros` and `ament_cmake_haros` are also needed in the `ament_lint` repo (sorry if this input comes kind of late - if we'd prefer this to be outside of `ament_lint` I'll take the effort to commit this into a separate repo).

Not an effort at all to move it into a different repo. I just put it to make sure it integrates well into the other ROS2 linters.

About the `FileNotFoundError`, sorry that is strange and should not happen.
I means that for some reason HAROS did not produce an XML output file.
I'll investigate it, but if you have time, can you please check if after running it you have a folder `/tmp/ament_haros/` and what it's contents are?

EDIT:
I tested it a bit more but could not reproduce the issue.
I have added an additional check if the XML file was written to make sure it doesn't fail silently, but please let me know if you have any idea why it might fail to do so on your system.

**Comment by nlimpert** [comment]
> About the `FileNotFoundError`, sorry that is strange and should not happen.
> I means that for some reason HAROS did not produce an XML output file.
> I'll investigate it, but if you have time, can you please check if after running it you have a folder `/tmp/ament_haros/` and what it's contents are?

`tree /tmp/ament_haros`
<details>
  <summary>Click to expand!</summary>

```
/tmp/ament_haros
├── ament_haros_project.yaml
├── haros_data
├── haros_home
├── haros-ros2-support-with-junitxml-improvements
│   ├── CHANGELOG.md
│   ├── docker
│   │   ├── Dockerfile
│   │   ├── haros_call.sh
│   │   ├── index.yaml
│   │   └── README.md
│   ├── docs
│   │   ├── css
│   │   │   ├── color.css
│   │   │   ├── style.css
│   │   │   └── type.css
│   │   ├── data
│   │   │   ├── haros_tutorials
│   │   │   │   ├── compliance
│   │   │   │   │   ├── runtime
│   │   │   │   │   │   ├── barebones.json
│   │   │   │   │   │   ├── double_remap.json
│   │   │   │   │   │   ├── dual_bots.json
│   │   │   │   │   │   ├── minimal.json
│   │   │   │   │   │   ├── missing_node.json
│   │   │   │   │   │   ├── missing_remap.json
│   │   │   │   │   │   ├── name_collision.json
│   │   │   │   │   │   ├── params.json
│   │   │   │   │   │   ├── rosparam.json
│   │   │   │   │   │   └── type_check.json
│   │   │   │   │   ├── source
│   │   │   │   │   │   ├── fictibot_controller.json
│   │   │   │   │   │   ├── fictibot_drivers.json
│   │   │   │   │   │   └── minimal_example.json
│   │   │   │   │   └── unknown.json
│   │   │   │   ├── configurations.json
│   │   │   │   ├── metrics
│   │   │   │   │   ├── fictibot_controller.json
│   │   │   │   │   ├── fictibot_drivers.json
│   │   │   │   │   └── minimal_example.json
│   │   │   │   ├── metrics.json
│   │   │   │   ├── packages.json
│   │   │   │   ├── rules.json
│   │   │   │   └── summary.json
│   │   │   └── projects.json
│   │   ├── index.html
│   │   ├── js
│   │   │   ├── app.js
│   │   │   ├── models.js
│   │   │   ├── routes.js
│   │   │   └── views
│   │   │       ├── dashboard.js
│   │   │       ├── general.js
│   │   │       ├── help.js
│   │   │       ├── issues.js
│   │   │       ├── packages.js
│   │   │       └── ros.js
│   │   ├── lib
│   │   │   ├── css
│   │   │   │   ├── font-awesome.css
│   │   │   │   ├── font-awesome.min.css
│   │   │   │   ├── grids-responsive-min.css
│   │   │   │   └── pure-min.css
│   │   │   ├── fonts
│   │   │   │   ├── FontAwesome.otf
│   │   │   │   ├── fontawesome-webfont.eot
│   │   │   │   ├── fontawesome-webfont.svg
│   │   │   │   ├── fontawesome-webfont.ttf
│   │   │   │   ├── fontawesome-webfont.woff
│   │   │   │   └── fontawesome-webfont.woff2
│   │   │   └── js
│   │   │       ├── backbone.min.js
│   │   │       ├── d3.min.js
│   │   │       ├── dagre.min.js
│   │   │       ├── jquery-2.2.1.min.js
│   │   │       └── underscore.min.js
│   │   └── screenshots
│   │       ├── dashboard_1.png
│   │       ├── dashboard_2.png
│   │       ├── issues.png
│   │       └── packages.png
│   ├── haros
│   │   ├── analysis_manager.py
│   │   ├── cmake_parser.py
│   │   ├── config_builder.py
│   │   ├── data.py
│   │   ├── data.pyc
│   │   ├── definitions.yaml
│   │   ├── export_manager.py
│   │   ├── extractor.py
│   │   ├── extractor.pyc
│   │   ├── haros.py
│   │   ├── haros.pyc
│   │   ├── __init__.py
│   │   ├── __init__.pyc
│   │   ├── launch_parser.py
│   │   ├── __main__.py
│   │   ├── metamodel.py
│   │   ├── metamodel.pyc
│   │   ├── models
│   │   │   ├── controller_manager.yaml
│   │   │   ├── diagnostic_aggregator.yaml
│   │   │   ├── hector_mapping.yaml
│   │   │   ├── husky_base.yaml
│   │   │   ├── interactive_marker_twist_server.yaml
│   │   │   ├── joy.yaml
│   │   │   ├── lms1xx.yaml
│   │   │   ├── move_base.yaml
│   │   │   ├── prosilica_camera.yaml
│   │   │   ├── robot_localization.yaml
│   │   │   ├── robot_state_publisher.yaml
│   │   │   ├── teleop_twist_joy.yaml
│   │   │   ├── template.yaml
│   │   │   └── tf.yaml
│   │   ├── plugin_manager.py
│   │   ├── pyflwor_monkey_patch.py
│   │   ├── util.py
│   │   └── visualiser.py
│   ├── haros-runner.py
│   ├── harosviz
│   │   ├── css
│   │   │   ├── color.css
│   │   │   ├── style.css
│   │   │   └── type.css
│   │   ├── index.html
│   │   ├── js
│   │   │   ├── app.js
│   │   │   ├── models.js
│   │   │   ├── routes.js
│   │   │   └── views
│   │   │       ├── dashboard.js
│   │   │       ├── general.js
│   │   │       ├── help.js
│   │   │       ├── issues.js
│   │   │       ├── packages.js
│   │   │       └── ros.js
│   │   └── lib
│   │       ├── css
│   │       │   ├── font-awesome.css
│   │       │   ├── font-awesome.min.css
│   │       │   ├── grids-responsive-min.css
│   │       │   └── pure-min.css
│   │       ├── fonts
│   │       │   ├── FontAwesome.otf
│   │       │   ├── fontawesome-webfont.eot
│   │       │   ├── fontawesome-webfont.svg
│   │       │   ├── fontawesome-webfont.ttf
│   │       │   ├── fontawesome-webfont.woff
│   │       │   └── fontawesome-webfont.woff2
│   │       └── js
│   │           ├── backbone.min.js
│   │           ├── d3.min.js
│   │           ├── dagre.min.js
│   │           ├── jquery-2.2.1.min.js
│   │           └── underscore.min.js
│   ├── LICENSE
│   ├── logo.png
│   ├── MODEL_EXTRACTION.md
│   ├── README.md
│   ├── README.rst
│   ├── requirements.txt
│   └── setup.py
└── haros_ros2-support.zip

29 directories, 133 files
```

</p>
</details>

---------------

I've just installed ROS Dashing in an Ubuntu 18.04 VM (to check whether the Fedora setup might be of reason for the failure). Currently I won't get a HAROS report either in this setup but the error message differs:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="ament_haros_test" tests="1" failures="1" time="0" errors="0" skip="0">
  <testcase classname="ament_haros_test" name="haros.xunit.missing_result" status="run" time="0">
    <failure message="The test did not generate a result file:&#10;&#10;HAROS failed to write xUnit (XML) output file&#10;"/>
  </testcase>
</testsuite>
```

Running `diff` on the two `tree` results in the following (note that I've piped the output of either tree to a filename matching the appropriate OS - e.g. /tmp/fedora_ament_haros being the one for Fedora):

`diff /tmp/fedora_ament_haros /tmp/ubuntu_ament_haros`
<details>
  <summary>Click to expand!</summary>

```
5c5,50
< ├── haros-ros2-support-with-junitxml-improvements
---
> │   ├── configs.yaml
> │   ├── export
> │   ├── index.yaml
> │   ├── parse_cache.json
> │   ├── projects
> │   │   └── default
> │   ├── pyflwor
> │   ├── repositories
> │   └── viz
> │       ├── css
> │       │   ├── color.css
> │       │   ├── style.css
> │       │   └── type.css
> │       ├── data
> │       ├── index.html
> │       ├── js
> │       │   ├── app.js
> │       │   ├── models.js
> │       │   ├── routes.js
> │       │   └── views
> │       │       ├── dashboard.js
> │       │       ├── general.js
> │       │       ├── help.js
> │       │       ├── issues.js
> │       │       ├── packages.js
> │       │       └── ros.js
> │       └── lib
> │           ├── css
> │           │   ├── font-awesome.css
> │           │   ├── font-awesome.min.css
> │           │   ├── grids-responsive-min.css
> │           │   └── pure-min.css
> │           ├── fonts
> │           │   ├── FontAwesome.otf
> │           │   ├── fontawesome-webfont.eot
> │           │   ├── fontawesome-webfont.svg
> │           │   ├── fontawesome-webfont.ttf
> │           │   ├── fontawesome-webfont.woff
> │           │   └── fontawesome-webfont.woff2
> │           └── js
> │               ├── backbone.min.js
> │               ├── d3.min.js
> │               ├── dagre.min.js
> │               ├── jquery-2.2.1.min.js
> │               └── underscore.min.js
> ├── haros-master
83a129
> │   │   ├── analysis_manager.pyc
84a131
> │   │   ├── cmake_parser.pyc
85a133
> │   │   ├── config_builder.pyc
89a138
> │   │   ├── export_manager.pyc
96a146
> │   │   ├── launch_parser.pyc
115a166
> │   │   ├── plugin_manager.pyc
118c169,171
< │   │   └── visualiser.py
---
> │   │   ├── util.pyc
> │   │   ├── visualiser.py
> │   │   └── visualiser.pyc
163c216
< └── haros_ros2-support.zip
---
> └── haros.zip
165c218
< 29 directories, 133 files
---
> 43 directories, 172 files
```

</p>
</details>

Yes I can have a look into this issue - IMO ideally there shouldn't be a difference between the two operating systems - both should work. I can have a look into this during the week.


BTW just to be complete - I've appended a `-b ament_haros` to the `git clone` command in https://github.com/rosin-project/haros_catkin/issues/24#issuecomment-531065325 because otherwise it won't have the correct branch - maybe you could update your comment.

**Comment by max-krichenbauer** [comment]
@nlimpert  Thanks for the test!

The differences in the error message (and file tree) are because I updated the `ament_haros` repo today. Sorry for the confusion.

It seems like HAROS silently fails on your system before it can generate any output.
My guess is a missing dependency.
Can you try `pip install haros haros-plugins`?
It won't use that version of HAROS because of the git download to `/tmp/` (until the pip version is updated), but I think pip should ensure that all requirements for haros are present.

**Comment by nlimpert** [comment]
Thanks for the note on installing `haros` and `haros-plugins` via `pip` - apparently this helped in terms of no errors occured. The content of `build/ament_haros_test/test_results/ament_haros_test/haros.xunit.xml` now looks like this:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<testsuites id="HAROS_project:ament_haros_project_2019-09-25-11-16" name="HAROS analysis result for project:ament_haros_project (2019-09-25-11-16)" tests="155" failures="0" time="0.075939">
  <testsuite id="HAROS.AnalysisReport.ament_haros_test" name="HAROS package analysis for ament_haros_test" tests="155" failures="0" time="0.075939">
  </testsuite>
</testsuites>
```

`/tmp/ament_haros/haros_data` seems to contain the actual HAROS output :+1:
Note that this was again tested on the Fedora 29 setup.

**Comment by max-krichenbauer** [comment]
Great! Thanks!
I'll update my comment above accordingly, but after all it's a temporary exception. Later, the package requirement should take care of that, right?

**Comment by nlimpert** [comment]
> Great! Thanks!
> I'll update my comment above accordingly, but after all it's a temporary exception. Later, the package requirement should take care of that, right?

Yes - actually if we're following the way of leaving the `ament_haros` and `ament_cmake_haros` in the `ament_lint` repo then we won't have to worry about this.
If, however, we consider those to be part of another repo we would have to follow the usual way of releasing the package (following the rules to do so). See [this PR](https://github.com/ros/rosdistro/pull/22146) for example.

-------

Looking at the [script ](https://github.com/esol-community/ament_lint/blob/c8d81f31d25fa45440b5edb033a6d04b588ec41a/ament_haros/ament_haros/main.py#L107) and your explanations I consider the intention is that `haros` is installed via pip, correct?
If so I think we might have to find a way to solve this similarly like `haros_catkin` currently does by using a `virtualenv` spawned by `catkin_virtualenv`. However I haven't found a similar functionality for `ament` yet.

**Comment by max-krichenbauer** [comment]
To be honest, I haven't yet fully grasped how the `haros` requirement is met in `haros_catkin` either.
I was going of @gavanderhoorn 's comment that

> So Haros would have to:
> either be a ROS package, or
> have a rosdep key
> For haros_catkin we've gone with the first option.

and just assumed it would eventually be possible to just add a dependency on `haros` in `package.xml` and `CMakeLists.txt`.

But I'll look into it now to get a better understanding of the process.

**Comment by nlimpert** [comment]
`haros_catkin` currently uses a [`requirements.txt`](https://github.com/rosin-project/haros_catkin/blob/d04963766f76bb29be732f7cf91a53920bdc1d9b/requirements.txt) which is read by [`catkin_virtualenv`](https://github.com/locusrobotics/catkin_virtualenv) which in turn creates a python virtualenv for `haros_catkin` and in there installs whatever is found in the `requirements.txt` via `pip`.

Note that the `requirements.txt` is supposed to be compatible to the [requirements files of pip](https://pip.readthedocs.io/en/1.1/requirements.html).

**Comment by max-krichenbauer** [comment]
I see. Thank you!
From looking at the code of `catkin_virtualenv` it seems like it really just creates a python virtualenv and installs the requested packages.
If there is no `ament` version of it, we can do it directly in `ament_haros`'s `main.py`.
Of course it "feels more clean" if we use another (officially sanctified) package, but the result should be the same, or am I mistaken?

**Comment by max-krichenbauer** [comment]
I found that for ROS2, providing 3rd-party software is commonly achieved through `_vendor` packages.

For example:
For [`libyaml`](https://github.com/yaml/libyaml) there is [`ros2/libyaml_vendor`](https://github.com/ros2/libyaml_vendor), a pure CMake package that will download, build and install `buildyaml`.
It is available as a [ROS2 package](https://index.ros.org/p/libyaml_vendor/#dashing) (repo [here](https://index.ros.org/r/libyaml_vendor/github-ros2-libyaml_vendor/) ).

Other (more complex) examples are:
[tinydir_vendor](https://github.com/ros2/tinydir_vendor)
[pybind11_vendor](https://github.com/ros2/pybind11_vendor)
[tinyxml_vendor](https://github.com/ros2/tinyxml_vendor)
[yaml_cpp_vendor](https://github.com/ros2/yaml_cpp_vendor)

Of course, in the end these also just download and install stuff which we could do directly in `ament_haros` (personally, I do not see a reason for making yet another package), but maybe this is considered "cleaner" or "following common procedures".

What do you think?

**Comment by nlimpert** [comment]
Thanks for the info on the `_vendor` packages - I wasn't aware of this.
If I understood correctly the `_vendor` packages seem to be used in order to fulfill dependencies that are to be installed via `CMake` - so I guess providing a virtualenv for such an environment would probably end up in an implementation of `caktin_virtualenv` for `ament` (as `catkin_virtualenv` seems to be `CMake` based).

I've also found [this discussion](https://answers.ros.org/question/332915/what-does-the-_vendor-suffix-mean-in-a-package-name/) - where nuclearsandwich describes it nicely; the main purpose of the `_vendor` packages seems to be "unless you find a system-installed version of the package, download it and compile yourself", which seems to apply to Windows mostly, where system dependencies can't be installed with a package manager - the explanation also points out the [`tinyxml_vendor`](https://github.com/ros2/tinyxml_vendor/blob/master/CMakeLists.txt) package which seems to describe the purpose - there's a casual [dependency on `tinyxml`](https://github.com/ros2/tinyxml_vendor/blob/3caa00bc33a5462e91eb5d44486ca3be4feaf62d/package.xml#L12) defined and if [this can't be found](https://github.com/ros2/tinyxml_vendor/blob/3caa00bc33a5462e91eb5d44486ca3be4feaf62d/CMakeLists.txt#L14) we'll tell `CMake` to [get a packages state](https://github.com/ros2/tinyxml_vendor/blob/3caa00bc33a5462e91eb5d44486ca3be4feaf62d/CMakeLists.txt#L18) and run `CMake` on it.
However as we're only using python I don't think we should follow this way. 

IMO the proper way would be to have a consistency in solving the dependency to `haros` in ROS and ROS2 (namely in a python virtualenv) but I also see that's probably not possible unless there's something like an `ament_virtualenv`.

However I guess it should be fairly straightforward to port the functionality of `catkin_virtualenv` to `ament` so if you and @gavanderhoorn agree we might create an issue on this at https://github.com/locusrobotics/catkin_virtualenv.

**Comment by max-krichenbauer** [comment]
Thanks for the explanation!
My experience in ROS development is too limited to understand which is the best way to do it.
(Personally, my instinct would be to just check and ensure that HAROS is installed directly in the package - adding *_vendor or *_virtualenv package seems like another layer of abstraction and complexity that can potentially fail).
I'll look into `catkin_virtualenv` and try to understand how it works.

By the way, wouldn't it be better to continue the discussion over at https://github.com/ament/ament_lint/issues/192 ?

**Comment by max-krichenbauer** [comment]
After looking into `catkin_virtualenv`, I'm worried that porting it to `ament` might involve a lot of work.

Firstly, `catkin` and `ament` are sufficiently different that there are almost no direct replacements for functions like `catkin_install_python`. In fact, in `ament` handling Python projects and handling `cmake` projects are separated by build type. So the `setup.py` also works quite differently.
(My understanding is still quite limited, but my guess is that this is the reason why the ROS2 linters are all split into two packages `ament_mylinter` and `ament_cmake_mylinter`).
So a rewrite of the build and install process seems necessary.

Secondly, `catkin_virtualenv` is using `catkin-pkg` for which there seems to be no `ament` counterpart. 
```
scripts/glob_requirements:27:from catkin_pkg.package import parse_package
setup.py:20:from catkin_pkg.python_setup import generate_distutils_setup
```
Furthermore,
```
./scripts/glob_requirements:26:from catkin.find_in_workspaces import find_in_workspaces
```
also seems to have no `ament` counterpart.

`catkin_virtualenv` also requires `rospkg` (not ROS2 compatible), but I could not immediately see where it is used anyways.

Seeing that everyone is busy with other projects as well and my understanding of the ROS build process is still quite limited, I'm wondering if this is really the most promising way.

Of course, I'd like to hear @gavanderhoorn 's opinion too.

**Comment by nlimpert** [comment]
Thanks for the investigation! I considered that maybe the maintainers of `catkin_virtualenv` might be interested in a port to `ament` as well so we might also create an issue on their repository, so they could join the discussion and collaborate.

Then they could also decide whether the port would be too much effort and/or there are other ways of solving this. What do you think?

**Comment by max-krichenbauer** [comment]
@nlimpert Yes, of course, we can ask. Do you want me to start the conversation?

**Comment by nlimpert** [comment]
> @nlimpert Yes, of course, we can ask. Do you want me to start the conversation?

That would be great as I think you're more familiar with the actual workflow of `catkin_virtualenv` than I am (at least because you investigated the workflow recently).
But maybe we can wait for @gavanderhoorn whether he thinks that we're going into the right direction in general.

**Comment by max-krichenbauer** [comment]
I started a discussion [here](https://github.com/locusrobotics/catkin_virtualenv/issues/47).

**Comment by max-krichenbauer** [comment]
@nlimpert 
I started working on an `ament` port of `catkin_virtualenv` over at https://github.com/esol-community/ament_virtualenv
I don't have a lot of confidence that it's working as intended. Feedback would be appreciated (over at https://github.com/locusrobotics/catkin_virtualenv/issues/47#issuecomment-537405186 )

**Comment by nlimpert** [comment]
Apologies for the late response - I haven't been able to look into your progress yet but your work looks very promising! I will try out your work as soon as possible.

**Comment by max-krichenbauer** [comment]
Update on the whole situation:
[ament_virtualenv](https://github.com/esol-community/ament_virtualenv) is finally available on rosdistro and the build farm, so the basic means for a linter package using HAROS can be satisfied.
[ament_haros and ament_cmake_haros](https://github.com/esol-community/ament_lint/tree/ament_haros) are also working,
but is still using a manual download of HAROS from github, because the pip release did not include the latest features.
Now I just saw that there is a new HAROS pip release as well (3.8.1), so the final step is to update `ament_haros` to us `ament_virtualenv` to get the new pip release of HAROS.

**Comment by max-krichenbauer** [comment]
Belated update:
[`ament_haros`](https://github.com/esol-community/ament_lint/tree/ament_haros) now properly uses `ament_virtualenv` and pip to use HAROS.
So I guess the development was completed and this issue can be closed.
